apps/sms/test/marionette/mocks/mock_navigator_moz_mobile_message.js

```diff
@@ -27,215 +32,240 @@ Services.obs.addObserver(function(document) {
-      receiver: parameters.receiver,
-      type: parameters.type,
-      delivery: 'sending',
-      deliveryInfo: [{
-        receiver: null,
-        deliveryStatus: 'not-applicable',
-        readStatus: 'not-applicable'
-      }],
-      subject: parameters.subject,
-      smil: parameters.smil,
-      body: parameters.body,
-      attachments: parameters.attachments,
-      timestamp: Date.now(),
-      sentTimestamp: Date.now(),
-      read: true
-    }, window));
-
-    thread.messages.push(message);
-
-    return message;
+      );
+
+      var message = {
+        id: ++storage.uniqueMessageIdCounter,
+        iccId: null,
+        threadId: thread.id,
+        sender: null,
+        receivers: parameters.receivers,
+        receiver: parameters.receiver,
+        type: parameters.type,
+        delivery: 'sending',
+        deliveryInfo: [{
+          receiver: null,
+          deliveryStatus: 'not-applicable',
+          readStatus: 'not-applicable'
+        }],
+        subject: parameters.subject,
+        smil: parameters.smil,
+        body: parameters.body,
+        attachments: parameters.attachments,
+        timestamp: Date.now(),
+        sentTimestamp: Date.now(),
+        read: true
+      };
+
+      thread.messages.push(message);
+
+      return message;
+    });
   }
 
   var MobileMessage = {
     getSegmentInfoForText: function(text) {
-      var request = Cu.waiveXrays(Cu.cloneInto({
-        onsuccess: null,
-        onerror: null
-      }, window));
-
-      window.setTimeout(function() {
-        var length = text.length;
-        var segmentLength = 160;
-        var charsUsedInLastSegment = (length % segmentLength);
-        var segments = Math.ceil(length / segmentLength);
-        if (typeof request.onsuccess === 'function') {
-          request.onsuccess.call(request, Cu.cloneInto({
-            target: {
-              result: {
-                segments: segments,
-                charsAvailableInLastSegment: charsUsedInLastSegment ?
-                  segmentLength - charsUsedInLastSegment : 0
-              }
-            }
-          }, window));
-        }
-      }, delayMs);
+      var request = Services.DOMRequest.createRequest(window);
+
+      var segmentLength = 160;
+      var charsUsedInLastSegment = (text.length % segmentLength);
+
+      // Very simple example of message segmentation logic, in reality it's more
+      // complex, e.g. see real "calculateLength" method here:
+      // https://android.googlesource.com/platform/frameworks/opt/telephony
+      // /+/android-5.0.1_r1/src/java/android/telephony/SmsMessage.java
+      var result = {
+        segments: Math.ceil(text.length / segmentLength),
+        charsAvailableInLastSegment: charsUsedInLastSegment ?
+          segmentLength - charsUsedInLastSegment : 0
+      };
+
+      // See description at the top of the file about "cloneInto" necessity.
+      Services.DOMRequest.fireSuccessAsync(
+        request, Cu.cloneInto(result, window)
```

Comment:
Nope, there is a ```Permission denied``` exception if I remove ```cloneInto```, does it work for you without ```cloneInto```? I believe it doesn't work for object, since content can remove/change properties of source object.